File: android/core/src/main/kotlin/cc/bsns/ssh/core/KnownHosts.kt
package cc.bsns.ssh.core

/** A server's host key as presented during the handshake. */
class HostKey(val keyType: String, val blob: ByteArray) {
    val fingerprint: String get() = SshKeyFormat.fingerprintOfPublicKeyBlob(blob)

    // Value equality (incl. blob bytes) to mirror the Swift Equatable HostKey.
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is HostKey) return false
        return keyType == other.keyType && blob.contentEquals(other.blob)
    }

    override fun hashCode(): Int = 31 * keyType.hashCode() + blob.contentHashCode()
}

/** Result of checking a presented host key against what we've stored. */
sealed interface HostVerification {
    /** Key matches a stored entry — proceed. */
    object Trusted : HostVerification
    /** No stored entry — first contact (TOFU). */
    data class Unknown(val fingerprint: String) : HostVerification
    /** A different key than stored — do not proceed without a loud override. */
    data class Mismatch(val stored: String, val presented: String) : HostVerification
}

/**
 * Trust-on-first-use host-key store — Kotlin port of the Swift `KnownHosts`.
 * Pure; the prompt UI and persistence live above it.
 */
class KnownHosts(entries: Map<String, HostKey> = emptyMap()) {
    private val entries = entries.toMutableMap()

    val allEntries: Map<String, HostKey> get() = entries.toMap()

    fun verify(host: String, port: Int, key: HostKey): HostVerification {
        val stored = entries[identifier(host, port)] ?: return HostVerification.Unknown(key.fingerprint)
        return if (stored.blob.contentEquals(key.blob)) HostVerification.Trusted
        else HostVerification.Mismatch(stored.fingerprint, key.fingerprint)
    }

    fun trust(host: String, port: Int, key: HostKey) {
        entries[identifier(host, port)] = key
    }

    fun storedKey(host: String, port: Int): HostKey? = entries[identifier(host, port)]

    fun forget(identifier: String) {
        entries.remove(identifier)
    }

    companion object {
        /** OpenSSH-style host identifier: bare host on port 22, else `[host]:port`. */
        fun identifier(host: String, port: Int): String = if (port == 22) host else "[$host]:$port"
    }
}

File: Sources/BsnsSSHCore/KnownHosts.swift
import Foundation

/// A server's host key as presented during the handshake.
public struct HostKey: Sendable, Equatable, Codable {
    /// SSH key-type name, e.g. "ssh-ed25519" / "ecdsa-sha2-nistp256".
    public let keyType: String
    /// Raw host-key blob (as returned by the transport).
    public let blob: Data

    public init(keyType: String, blob: Data) {
        self.keyType = keyType
        self.blob = blob
    }

    public var fingerprint: String { SSHKeyFormat.fingerprint(ofPublicKeyBlob: blob) }
}

/// The result of checking a presented host key against what we've stored.
public enum HostVerification: Sendable, Equatable {
    /// Key matches a stored entry — proceed.
    case trusted
    /// No stored entry — first contact (TOFU). Prompt the user with the
    /// fingerprint; on accept, call `trust(...)`.
    case unknown(fingerprint: String)
    /// A different key than the one stored — the dangerous case. Do NOT
    /// proceed without an explicit, loud user override.
    case mismatch(stored: String, presented: String)
}

/// Trust-on-first-use host-key store. Pure + serializable; the prompt UI and
/// persistence (sync blob / disk) live above it.
public struct KnownHosts: Sendable, Equatable, Codable {
    private var entries: [String: HostKey]

    public init(entries: [String: HostKey] = [:]) {
        self.entries = entries
    }

    /// Check a presented key. Never mutates — accepting a TOFU prompt is an
    /// explicit `trust(...)` call by the caller.
    public func verify(host: String, port: UInt16, key: HostKey) -> HostVerification {
        guard let stored = entries[Self.identifier(host, port)] else {
            return .unknown(fingerprint: key.fingerprint)
        }
        if stored.blob == key.blob { return .trusted }
        return .mismatch(stored: stored.fingerprint, presented: key.fingerprint)
    }

    /// Record a key as trusted for a host (after a TOFU accept, or an explicit
    /// override on mismatch).
    public mutating func trust(host: String, port: UInt16, key: HostKey) {
        entries[Self.identifier(host, port)] = key
    }

    public func storedKey(host: String, port: UInt16) -> HostKey? {
        entries[Self.identifier(host, port)]
    }

    public var allEntries: [String: HostKey] { entries }

    /// Forget a trusted host by its identifier (a key from `allEntries`).
    public mutating func forget(_ identifier: String) {
        entries[identifier] = nil
    }

    /// OpenSSH-style host identifier: bare host on port 22, else `[host]:port`.
    static func identifier(_ host: String, _ port: UInt16) -> String {
        port == 22 ? host : "[\(host)]:\(port)"
    }
}
